src/stream/src/executor/sort.rs

```diff
@@ -0,0 +1,506 @@
+    pub fn new(
+        context: ActorContextRef,
+        input: BoxedExecutor,
+        pk_indices: PkIndices,
+        executor_id: u64,
+        state_table: StateTable<S>,
+        sort_column_index: usize,
+    ) -> Self {
+        let schema = input.schema().clone();
+        Self {
+            context,
+            input: Some(input),
+            pk_indices,
+            identity: format!("SortExecutor {:X}", executor_id),
+            schema,
+            state_table,
+            sort_column_index,
+            buffer: BTreeMap::new(),
+            _prev_watermark: None,
+        }
+    }
+
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    async fn execute_inner(mut self) {
+        let mut input = self.input.take().unwrap().execute();
+
+        // Consume the first barrier message and initialize state table.
+        let barrier = expect_first_barrier(&mut input).await?;
+        self.state_table.init_epoch(barrier.epoch);
+
+        // The first barrier message should be propagated.
+        yield Message::Barrier(barrier);
+
+        // Fill the buffer on initialization. If the executor has recovered from a failed state, the
+        // buffer should be refilled by its previous data.
+        let vnode_bitmap = self.state_table.vnode_bitmap().to_owned();
+        self.fill_buffer(None, &vnode_bitmap).await;
+
+        #[for_await]
+        for msg in input {
+            match msg? {
+                Message::Watermark(watermark) => {
+                    let Watermark { col_idx, val } = watermark.clone();
+                    // Sort executor only sends a stream chunk to downstream when
+                    // `self.sort_column_index` matches the watermark's column index. Otherwise, it
+                    // just forwards the watermark message to downstream without sending a stream
+                    // chunk message.
+                    if col_idx == self.sort_column_index {
+                        let mut stream_chunk_data = Vec::with_capacity(self.buffer.len());
+                        let watermark_value = val.clone();
+
+                        // Find out the records to send to downstream.
+                        while let Some(entry) = self.buffer.first_entry() {
+                            // Only when a record's timestamp is prior to or equivalent to the
+                            // watermark should it be sent to downstream.
+                            if entry.key().0 <= watermark_value {
+                                // Remove the record from memory.
+                                let (op, row, persisted) = entry.remove();
+                                // Remove the record from state store. It is possible that a record
+                                // is not present in state store because this watermark arrives
+                                // before a barrier since last watermark.
+                                // TODO: Use range delete instead.
```

Comment:
Actually, it's okay to write some extra tombstones to the storage, but let's keep this to make the sanity check happy. 🤣